nexus/src/app/instance.rs

```diff
@@ -72,48 +74,90 @@ type SledAgentClientError =
 
-impl std::fmt::Display for SledAgentInstancePutError {
+impl std::fmt::Display for SledAgentInstanceError {
     fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
         write!(f, "{}", self.0)
     }
 }
 
-impl From<SledAgentClientError> for SledAgentInstancePutError {
+impl From<SledAgentClientError> for SledAgentInstanceError {
     fn from(value: SledAgentClientError) -> Self {
         Self(value)
     }
 }
 
-impl From<SledAgentInstancePutError> for omicron_common::api::external::Error {
-    fn from(value: SledAgentInstancePutError) -> Self {
+impl From<SledAgentInstanceError> for omicron_common::api::external::Error {
+    fn from(value: SledAgentInstanceError) -> Self {
         value.0.into()
     }
 }
 
-impl SledAgentInstancePutError {
-    /// Returns `true` if this error is of a class that indicates that Nexus
-    /// cannot assume anything about the health of the instance or its sled.
+impl From<SledAgentInstanceError> for dropshot::HttpError {
+    fn from(value: SledAgentInstanceError) -> Self {
+        use dropshot::HttpError;
+        use progenitor_client::Error as ClientError;
+        // See RFD 486 §6.3:
+        // https://rfd.shared.oxide.computer/rfd/486#_stopping_or_rebooting_a_running_instance
+        match value {
+            // Errors communicating with the sled-agent should return 502 Bad
+            // Gateway to the caller.
+            SledAgentInstanceError(ClientError::CommunicationError(e)) => {
+                // N.B.: it sure *seems* like we should be using the
+                // `HttpError::for_status` constructor here, but that
+                // constructor secretly calls the `for_client_error`
+                // constructor, which panics if we pass a status code that isn't
+                // a 4xx error. So, instead, we construct an internal error and
+                // then munge its status code.
+                let mut error = HttpError::for_internal_error(e.to_string());
+                error.status_code = if e.is_timeout() {
+                    // This isn't actually in the RFD, but I thought it might be
+                    // nice to return 504 Gateway Timeout when the communication
+                    // error is a timeout...
+                    http::StatusCode::GATEWAY_TIMEOUT
+                } else {
+                    http::StatusCode::BAD_GATEWAY
+                };
+                error
+            }
+            // Invalid request errors from the sled-agent should return 500
+            // Internal Server Errors.
+            SledAgentInstanceError(ClientError::InvalidRequest(s)) => {
+                HttpError::for_internal_error(s)
+            }
+            // Error responses from sled-agent that indicate the instance is
+            // unhealthy should be mapped to a 500 error.
+            e if e.instance_unhealthy() => {
+                HttpError::for_internal_error(e.to_string())
```

Comment:
Yeah, that seems like a pretty good message. I'm not sure whether it ought to be a 503 or a 500 off the top of my head, though.

Reply:
My mental model is that it's useful to use 500 essentially to denote an internal assertion failure or failed invariant. To me, the nice thing about that is that `omicron-stress` can then log/halt testing/whatever for all 500s, since none of them are ever expected operational conditions. But this is just one person's opinion--there might be a design doc or RFD contradicting it somewhere.